File: everyDay/1792.最大平均通过率.cpp
/*
 * @lc app=leetcode.cn id=1792 lang=cpp
 *
 * [1792] 最大平均通过率
 */

// @lc code=start
class Solution {
public:
    struct node {
        int p, t;
        bool operator < ( const node &n) const {
            return (long long) n.t *(n.t + 1) * (t - p) < (long long) t * (t + 1) * (n.t - n.p);
        }
    };
    double maxAverageRatio(vector<vector<int>>& classes, int extraStudents) {
/**
p = (pi + x) / (ti + x) + ...
2/ 4 - 3/ 5
pi * (ti + 1) - (pi + 1) * ti
-----------------------------
ti * (ti + 1)
==>

维护一个按增量最大的优先队列

ti - pi
-------
ti * (ti + 1) 


*/
        priority_queue<node> P;
        for (auto &c : classes) {
            P.push({c[0], c[1]});
        }
        for (int i = 0; i < extraStudents; i++) {
            auto [pass, total] = P.top();
            P.pop();
            P.push({pass + 1, total + 1});
        }
        double ans = 0;
        for (int i = 0; i < classes.size(); i++) {
            auto [pass, total] = P.top();
            P.pop();
            ans += 1.0 * pass / total;
        }
        return ans / classes.size();
    }
};
// @lc code=end

File: everyDay/1792.最大平均通过率.go
/*
 * @lc app=leetcode.cn id=1792 lang=golang
 *
 * [1792] 最大平均通过率
 */

// @lc code=start

func maxAverageRatio(classes [][]int, extraStudents int) (ans float64) {
    h := hp(classes)
    heap.Init(&h)
    for ; extraStudents > 0; extraStudents-- {
        h[0][0]++
        h[0][1]++
        heap.Fix(&h, 0)
    }
    for _, c := range h {
        ans += float64(c[0]) / float64(c[1])
    }
    return ans / float64(len(classes))
}

type hp [][]int
func (h hp) Len() int { return len(h) }
func (h hp) Less(i, j int) bool { a, b := h[i], h[j]; return (a[1]-a[0])*b[1]*(b[1]+1) > (b[1]-b[0])*a[1]*(a[1]+1) }
func (h hp) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (hp) Push(interface{})     {}
func (hp) Pop() (_ interface{}) { return }
// @lc code=end

